### DISCUSSION TITLE
How do I pip install a package in a private repo?

### DISCUSSION THREAD

**Discussion by kcpevey** [nebari-dev/nebari#283] [Q&A] [answered]
I have a package in a private repo on github. Locally, I would do a pip install of this package. How can I do this on qhub? 

It seems that via the qhub_config.yaml it would need to be public?

**Comment by costrouc** [comment, marked as answer]
There are two common approaches both use the fact that pip can install from a url a tarball. https://pip.pypa.io/en/stable/reference/pip_install/#git. If you use the `git+ssh` you need to setup an [github deploy key](https://docs.github.com/en/developers/overview/managing-deploy-keys) and setup your ssh `.ssh/config` to use the correct ssh private key. Or you can embed the username and api token https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token. You would call it as `git+https://<username>:<token>@:git.example.com/MyProject#egg=MyProject`

**Reply by kcpevey** [reply]
Thanks for explaining!

**Comment by Adam-D-Lewis** [comment]
I'm closing old discussions in a repo cleanup effort.  Feel free to re-open if needed.

**Comment by kcpevey** [comment]
We also now have this explained in the docs - https://www.nebari.dev/docs/how-tos/install-pip-packages#installing-packages-from-a-private-github-repository
